File: BotsSDK/korebot/src/main/java/kore/botssdk/dialogs/ResetPinTemplateBottomSheet.java
package kore.botssdk.dialogs;

import static android.content.Context.MODE_PRIVATE;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static kore.botssdk.models.BotResponse.BUBBLE_RIGHT_BG_COLOR;
import static kore.botssdk.models.BotResponse.BUBBLE_RIGHT_TEXT_COLOR;
import static kore.botssdk.models.BotResponse.THEME_NAME;
import static kore.botssdk.view.viewUtils.DimensionUtil.dp1;
import static kore.botssdk.viewholders.BaseViewHolder.getDotMessage;
import static kore.botssdk.viewholders.BaseViewHolder.setRoundedCorner;

import android.app.Dialog;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import kore.botssdk.R;
import kore.botssdk.adapter.PinFieldItemAdapter;
import kore.botssdk.application.AppControl;
import kore.botssdk.itemdecoration.HorizontalSpaceItemDecoration;
import kore.botssdk.listener.ComposeFooterInterface;
import kore.botssdk.models.PayloadInner;
import kore.botssdk.utils.StringUtils;

public class ResetPinTemplateBottomSheet extends BottomSheetDialogFragment {
    private ComposeFooterInterface composeFooterInterface;
    private BottomSheetDialog bottomSheetDialog;
    private PayloadInner payloadInner;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.reset_pin_template_bottom_sheet, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        AppCompatImageView ivClose = view.findViewById(R.id.close);
        TextView tvTitle = view.findViewById(R.id.title);
        TextView tvEnterNewPin = view.findViewById(R.id.enter_new_pin);
        TextView tvRenterNewPin = view.findViewById(R.id.reenter_new_pin);
        RecyclerView newPinRecycler = view.findViewById(R.id.new_pin_recycler);
        RecyclerView reenterPinRecycler = view.findViewById(R.id.reenter_pin_recycler);
        TextView tvReset = view.findViewById(R.id.reset);
        newPinRecycler.addItemDecoration(new HorizontalSpaceItemDecoration(10));
        newPinRecycler.setLayoutManager(new LinearLayoutManager(view.getContext(), LinearLayoutManager.HORIZONTAL, false));
        newPinRecycler.post(() -> {
            newPinRecycler.setAdapter(new PinFieldItemAdapter(payloadInner.getPinLength(), newPinRecycler.getMeasuredWidth()));
            newPinRecycler.post(() -> {
                int childCount = newPinRecycler.getChildCount();

                for (int index = 0; index < childCount; index++) {
                    EditText childView = newPinRecycler.getChildAt(index).findViewById(R.id.otp_field);

                    EditText prevChildView = (index - 1 >= 0)
                            ? newPinRecycler.getChildAt(index - 1).findViewById(R.id.otp_field)
                            : null;

                    EditText nextChildView = (index + 1 < childCount)
                            ? newPinRecycler.getChildAt(index + 1).findViewById(R.id.otp_field)
                            : null;

                    addTextWatcher(prevChildView, childView, nextChildView);
                }
            });
        });

        reenterPinRecycler.addItemDecoration(new HorizontalSpaceItemDecoration(10));
        reenterPinRecycler.setLayoutManager(new LinearLayoutManager(view.getContext(), LinearLayoutManager.HORIZONTAL, false));
        reenterPinRecycler.post(() -> {
            reenterPinRecycler.setAdapter(new PinFieldItemAdapter(payloadInner.getPinLength(), reenterPinRecycler.getMeasuredWidth()));
            reenterPinRecycler.post(() -> {
                int childCount = reenterPinRecycler.getChildCount();

                for (int index = 0; index < childCount; index++) {
                    EditText childView = reenterPinRecycler.getChildAt(index).findViewById(R.id.otp_field);

                    EditText prevChildView = (index - 1 >= 0)
                            ? reenterPinRecycler.getChildAt(index - 1).findViewById(R.id.otp_field)
                            : null;

                    EditText nextChildView = (index + 1 < childCount)
                            ? reenterPinRecycler.getChildAt(index + 1).findViewById(R.id.otp_field)
                            : null;

                    addTextWatcher(prevChildView, childView, nextChildView);
                }
            });
        });

        tvTitle.setVisibility(!StringUtils.isNullOrEmpty(payloadInner.getTitle()) ? VISIBLE : GONE);
        tvTitle.setText(payloadInner.getTitle());
        tvEnterNewPin.setText(payloadInner.getEnterPinTitle());
        tvRenterNewPin.setText(payloadInner.getReEnterPinTitle());

        if (payloadInner.getResetButtons() != null && !payloadInner.getResetButtons().isEmpty()) {
            tvReset.setText(payloadInner.getResetButtons().get(0).getTitle());
            SharedPreferences sharedPrefs = view.getContext().getSharedPreferences(THEME_NAME, MODE_PRIVATE);
            String bgColor = sharedPrefs.getString(BUBBLE_RIGHT_BG_COLOR, "#3F51B5");
            String txtColor = sharedPrefs.getString(BUBBLE_RIGHT_TEXT_COLOR, "#FFFFFF");
            setRoundedCorner(tvReset, 6 * dp1);
            tvReset.setBackgroundColor(Color.parseColor(bgColor));
            tvReset.setTextColor(Color.parseColor(txtColor));
            tvReset.setOnClickListener(v -> {
                StringBuilder enteredPinBuilder = new StringBuilder();
                int childCount = newPinRecycler.getChildCount();

                for (int index = 0; index < childCount; index++) {
                    EditText childView = newPinRecycler.getChildAt(index).findViewById(R.id.otp_field);
                    enteredPinBuilder.append(childView.getText().toString());
                }
                String enteredPin = enteredPinBuilder.toString();

                StringBuilder reenteredPinBuilder = new StringBuilder();
                childCount = reenterPinRecycler.getChildCount();

                for (int index = 0; index < childCount; index++) {
                    EditText childView = reenterPinRecycler.getChildAt(index).findViewById(R.id.otp_field);
                    reenteredPinBuilder.append(childView.getText().toString());
                }
                String reenteredPin = reenteredPinBuilder.toString();

                if (enteredPin.length() == payloadInner.getPinLength() && reenteredPin.length() == payloadInner.getPinLength() && enteredPin.equals(reenteredPin)) {
                    if (composeFooterInterface != null && enteredPin.length() == payloadInner.getPinLength()) {
                        composeFooterInterface.onSendClick(getDotMessage(enteredPin), enteredPin, false);
                        if (bottomSheetDialog != null) bottomSheetDialog.dismiss();
                    }
                } else {
                    Toast.makeText(view.getContext(), payloadInner.getWarningMessage(), Toast.LENGTH_LONG).show();
                }
            });
        }

        ivClose.setOnClickListener(v -> {
            if (bottomSheetDialog != null) bottomSheetDialog.dismiss();
        });
    }

    public void setData(PayloadInner payloadInner) {
        this.payloadInner = payloadInner;
    }

    public void setComposeFooterInterface(ComposeFooterInterface composeFooterInterface) {
        this.composeFooterInterface = composeFooterInterface;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        bottomSheetDialog = (BottomSheetDialog) super.onCreateDialog(savedInstanceState);
        bottomSheetDialog.setOnShowListener(dialogInterface -> {
            BottomSheetDialog d = (BottomSheetDialog) dialogInterface;
            FrameLayout bottomSheet = d.findViewById(com.google.android.material.R.id.design_bottom_sheet);

            if (bottomSheet != null) {
                if (bottomSheet.getLayoutParams() != null)
                    bottomSheet.getLayoutParams().height = (int) ((AppControl.getInstance(getContext()).getDimensionUtil().screenHeight) - (50 * dp1));

                BottomSheetBehavior bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet);
                bottomSheetBehavior.setPeekHeight((int) ((AppControl.getInstance(getContext()).getDimensionUtil().screenHeight) - (50 * dp1)));
            }
        });

        return bottomSheetDialog;
    }

    private void addTextWatcher(EditText previousView, EditText editText, EditText nextView) {
        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // no-op
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // no-op
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (!editText.getText().toString().trim().isEmpty()) {
                    if (nextView != null) {
                        nextView.requestFocus();
                    }
                } else {
                    if (previousView != null) {
                        previousView.requestFocus();
                    }
                }
            }
        });
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/bottomsheet/ResetPinTemplateBottomSheet.kt
package com.kore.ui.bottomsheet

import android.app.Dialog
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.Toast
import androidx.core.view.isVisible
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.FragmentManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.kore.common.event.UserActionEvent
import com.kore.event.BotChatEvent
import com.kore.extensions.dpToPx
import com.kore.extensions.getDotMessage
import com.kore.model.constants.BotResponseConstants.ENTER_PIN_TITLE
import com.kore.model.constants.BotResponseConstants.KEY_TITLE
import com.kore.model.constants.BotResponseConstants.PIN_LENGTH
import com.kore.model.constants.BotResponseConstants.REENTER_PIN_TITLE
import com.kore.model.constants.BotResponseConstants.RESET_BUTTONS
import com.kore.model.constants.BotResponseConstants.WARNING_MESSAGE
import com.kore.ui.R
import com.kore.ui.bottomsheet.row.PinFieldItemRowType
import com.kore.ui.bottomsheet.row.pinfielditem.PinFieldItemRow.Companion.createRows
import com.kore.ui.databinding.ResetPinTemplateBottomSheetBinding
import com.kore.ui.row.SimpleListAdapter
import com.kore.ui.row.SimpleListRow.HorizontalSpaceItemDecoration

class ResetPinTemplateBottomSheet : BottomSheetDialogFragment() {
    private var binding: ResetPinTemplateBottomSheetBinding? = null
    private var bottomSheetDialog: BottomSheetDialog? = null
    private lateinit var payload: Map<String, Any>
    private var isLastItem: Boolean = false
    private lateinit var actionEvent: (event: UserActionEvent) -> Unit
    private val itemSpace = 10

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.reset_pin_template_bottom_sheet, container, false)
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding?.apply {
            template.title.isVisible = payload[KEY_TITLE] != null
            template.title.text = payload[KEY_TITLE] as String?
            template.enterNewPin.text = payload[ENTER_PIN_TITLE] as String?
            template.reenterNewPin.text = payload[REENTER_PIN_TITLE] as String?

            val pinLength = (payload[PIN_LENGTH] as Double).toInt()
            template.newPinRecycler.addItemDecoration(HorizontalSpaceItemDecoration(itemSpace))
            template.newPinRecycler.layoutManager = LinearLayoutManager(root.context, LinearLayoutManager.HORIZONTAL, false)
            template.newPinRecycler.post {
                template.newPinRecycler.adapter = SimpleListAdapter(PinFieldItemRowType.values().asList()).apply {
                    submitList(createRows(pinLength, template.newPinRecycler.width, itemSpace))
                    template.newPinRecycler.post {
                        val childCount = template.newPinRecycler.childCount
                        for (index in 0 until childCount) {
                            val childView: EditText = template.newPinRecycler.getChildAt(index).findViewById(R.id.otp_field)
                            val prevChildView: EditText? =
                                if (index - 1 >= 0) template.newPinRecycler.getChildAt(index - 1).findViewById(R.id.otp_field) else null
                            val nextChildView: EditText? =
                                if (index + 1 < childCount) template.newPinRecycler.getChildAt(index + 1).findViewById(R.id.otp_field) else null
                            addTextWatcher(prevChildView, childView, nextChildView)
                        }
                    }
                }
            }

            template.reenterPinRecycler.addItemDecoration(HorizontalSpaceItemDecoration(itemSpace))
            template.reenterPinRecycler.layoutManager = LinearLayoutManager(root.context, LinearLayoutManager.HORIZONTAL, false)
            template.reenterPinRecycler.post {
                template.reenterPinRecycler.adapter = SimpleListAdapter(PinFieldItemRowType.values().asList()).apply {
                    submitList(createRows(pinLength, template.reenterPinRecycler.width, itemSpace))
                    template.reenterPinRecycler.post {
                        val childCount = template.reenterPinRecycler.childCount
                        for (index in 0 until childCount) {
                            val childView: EditText = template.reenterPinRecycler.getChildAt(index).findViewById(R.id.otp_field)
                            val prevChildView: EditText? =
                                if (index - 1 >= 0) template.reenterPinRecycler.getChildAt(index - 1).findViewById(R.id.otp_field) else null
                            val nextChildView: EditText? =
                                if (index + 1 < childCount) template.reenterPinRecycler.getChildAt(index + 1).findViewById(R.id.otp_field) else null
                            addTextWatcher(prevChildView, childView, nextChildView)
                        }
                    }
                }
            }

            (payload[RESET_BUTTONS] as List<Map<String, String>>?)?.let { list ->
                template.reset.text = list[0][KEY_TITLE]
                template.reset.setOnClickListener {
                    var enteredPin = ""
                    var childCount = template.newPinRecycler.childCount
                    for (index in 0 until childCount) {
                        val childView: EditText = template.newPinRecycler.getChildAt(index).findViewById(R.id.otp_field)
                        enteredPin += childView.text.toString()
                    }
                    var reenteredPin = ""
                    childCount = template.reenterPinRecycler.childCount
                    for (index in 0 until childCount) {
                        val childView: EditText = template.reenterPinRecycler.getChildAt(index).findViewById(R.id.otp_field)
                        reenteredPin += childView.text.toString()
                    }
                    if (enteredPin.length == pinLength && reenteredPin.length == pinLength && enteredPin == reenteredPin) {
                        bottomSheetDialog?.dismiss()
                        actionEvent(BotChatEvent.SendMessage(enteredPin.getDotMessage(), enteredPin))
                    } else {
                        Toast.makeText(root.context, payload[WARNING_MESSAGE].toString(), Toast.LENGTH_LONG).show()
                    }
                }
            }
            template.close.setOnClickListener {
                bottomSheetDialog?.dismiss()
            }
        }
    }

    private fun addTextWatcher(previousView: EditText?, editText: EditText, nextView: EditText?) {
        editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable?) {
                if (editText.text.isNotEmpty()) {
                    nextView?.requestFocus()
                } else {
                    previousView?.requestFocus()
                }
            }
        })
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        bottomSheetDialog = super.onCreateDialog(savedInstanceState) as BottomSheetDialog
        bottomSheetDialog?.setOnShowListener { dialogInterface ->
            val d = dialogInterface as BottomSheetDialog
            val bottomSheet = d.findViewById<FrameLayout>(com.google.android.material.R.id.design_bottom_sheet)
            if (bottomSheet != null) {
                bottomSheet.layoutParams.height = resources.displayMetrics.heightPixels
                val bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet)
                bottomSheetBehavior.peekHeight = 400.dpToPx(requireContext())
            }
        }

        return bottomSheetDialog!!
    }

    fun showData(
        payload: Map<String, Any>,
        isLastItem: Boolean,
        manager: FragmentManager,
        actionEvent: (event: UserActionEvent) -> Unit
    ) {
        this.payload = payload
        this.isLastItem = isLastItem
        this.actionEvent = actionEvent
        show(manager, ResetPinTemplateBottomSheet::class.java.name)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}